br/pkg/lightning/backend/external/writer.go

```diff
@@ -0,0 +1,340 @@
+	b.writeBatchCount = count
+	return b
+}
+
+// SetPropSizeDistance sets the distance of range size for each property.
+func (b *WriterBuilder) SetPropSizeDistance(dist uint64) *WriterBuilder {
+	b.propSizeDist = dist
+	return b
+}
+
+// SetPropKeysDistance sets the distance of range keys for each property.
+func (b *WriterBuilder) SetPropKeysDistance(dist uint64) *WriterBuilder {
+	b.propKeysDist = dist
+	return b
+}
+
+// SetOnCloseFunc sets the callback function when a writer is closed.
+func (b *WriterBuilder) SetOnCloseFunc(onClose OnCloseFunc) *WriterBuilder {
+	b.onClose = onClose
+	return b
+}
+
+// SetBufferPool sets the buffer pool of the writer.
+func (b *WriterBuilder) SetBufferPool(bufferPool *membuf.Pool) *WriterBuilder {
+	b.bufferPool = bufferPool
+	return b
+}
+
+// Build builds a new Writer.
+func (b *WriterBuilder) Build(
+	store storage.ExternalStorage,
+	writerID int,
+	filenamePrefix string,
+) *Writer {
+	bp := b.bufferPool
+	if bp == nil {
+		bp = membuf.NewPool()
+	}
+	return &Writer{
+		rc: &rangePropertiesCollector{
+			props:        make([]*rangeProperty, 0, 1024),
+			currProp:     &rangeProperty{},
+			propSizeDist: b.propSizeDist,
+			propKeysDist: b.propKeysDist,
+		},
+		memSizeLimit:   b.memSizeLimit,
+		store:          store,
+		kvBuffer:       bp.NewBuffer(),
+		writeBatch:     make([]common.KvPair, 0, b.writeBatchCount),
+		currentSeq:     0,
+		filenamePrefix: filenamePrefix,
+		writerID:       writerID,
+		kvStore:        nil,
+		onClose:        b.onClose,
+		closed:         false,
+	}
+}
+
+// Writer is used to write data into external storage.
+type Writer struct {
+	store          storage.ExternalStorage
+	writerID       int
+	currentSeq     int
+	filenamePrefix string
+
+	kvStore *KeyValueStore
+	rc      *rangePropertiesCollector
+
+	memSizeLimit uint64
+
+	kvBuffer   *membuf.Buffer
+	writeBatch []common.KvPair
+
+	onClose OnCloseFunc
+	closed  bool
+
+	// Statistic information per batch.
+	batchSize uint64
+
+	// Statistic information per writer.
+	minKey    tidbkv.Key
+	maxKey    tidbkv.Key
+	totalSize uint64
+}
+
+// AppendRows appends rows to the external storage.
+// Note that this method is NOT thread-safe.
+func (w *Writer) AppendRows(ctx context.Context, _ []string, rows encode.Rows) error {
+	kvs := kv.Rows2KvPairs(rows)
+	for _, pair := range kvs {
+		w.batchSize += uint64(len(pair.Key) + len(pair.Val))
+		key := w.kvBuffer.AddBytes(pair.Key)
+		val := w.kvBuffer.AddBytes(pair.Val)
+		w.writeBatch = append(w.writeBatch, common.KvPair{Key: key, Val: val})
+		if w.batchSize >= w.memSizeLimit {
+			if err := w.flushKVs(ctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// IsSynced implements the backend.EngineWriter interface.
+func (w *Writer) IsSynced() bool {
+	return false
+}
+
+// Close closes the writer.
+func (w *Writer) Close(ctx context.Context) (backend.ChunkFlushStatus, error) {
+	if w.closed {
+		return status(true), nil
```

Comment:
What's the situation w.closed is true?

Reply:
Normally, it should be closed only once.

Reply:
So, if it's a bug that w.closed is true, we can print an error log.